19.0 tutorial roole

roole-odoo opened the pull request:
# [roole] chapter 2

Finished Chapter 2 for roole.

roole-odoo committed:
[ADD] estate: Init the module

roole-odoo committed:
[IMP] estate: manifest changed
Changed the manifest to test the 'git commit' env

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1026.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1026)

roole-odoo committed:
[IMP] estate: Chapter 3

Chapter 3 finished.

roole-odoo committed:
[IMP] estate: Chapter 3 security

- the ir.model.access.csv have been created
- read, write, create and unlink permissions have been given to the group base.group_user.

roole-odoo committed:
[FIX] estate: Fixing for the Green box

Fixing manifest with license

roole-odoo committed:
[IMP] estate : Chapter 5

- an action have been loaded in the system.
- Property view created.
- view Menu added.
- Extra attributes have been added.

roole-odoo committed:
[FIX] estate: PEP8 correction

clbr-odoo commented on estate/models/estate_model.py:
We prevent from appending `_model` on every model. Since your model is `estate.property` it should be `estate_property.py`

clbr-odoo commented on estate/models/estate_model.py:
```diff
@@ -0,0 +1,33 @@
+from odoo import fields, models
+
+
+class EstateModel(models.Model):
```
```suggestion
class EstateProperty(models.Model):
```
Following previous comment.

clbr-odoo commented on estate/models/estate_model.py:
```diff
@@ -0,0 +1,33 @@
+class EstateModel(models.Model):
+    _name = "estate.property"
+    _description = "Estate description"
+    _property = "property"
```
```suggestion
```
This is doing nothing, remove. Also add an empty line between "class metadata" and the fields description.

clbr-odoo commented on estate/models/estate_model.py:
```diff
@@ -0,0 +1,33 @@
+    name = fields.Char(required=True)
+    description = fields.Text()
+    postcode = fields.Char()
+    date_availability = fields.Date(copy=False, default=datetime.datetime.now() + datetime.timedelta(days=90))
```
```suggestion
    date_availability = fields.Date(
        copy=False, 
        default=fields.Date.add(fields.Date.today(), months=3),
        string="Available from",
    )
```
1. Using datetime/relativedelta is not wrong (more or less, see next comment) but just so you know: there are helper that exists within Odoo for this, which avoids a new import.
2. Your field is a Date type, not Datetime, therefore using datetime.now() adds a cast that is easily avoidable.
3. We usually use the biggest unit possible: `days` -> `months`.
4. Also FYI, if not provided though the `string` parameter, Odoo deduce a label to display in the UI. In your case, it will deduce "Date Availability", which is not very user friendly 😉

clbr-odoo commented on estate/models/estate_model.py:
```diff
@@ -0,0 +1,33 @@
+    active = fields.Boolean(default=True)
+    state = fields.Selection(
+        string="Estate status",
+        selection=[('new', 'New'), ('offer received', 'Offer Received'), ('offer accepted', 'Offer Accepted'), ('sold', 'Sold'), ('cancelled', 'Cancelled')],
```
```suggestion
        selection=[
            ('new', 'New'), 
            ('offer_received', 'Offer Received'), 
            ('offer_accepted', 'Offer Accepted'), 
            ('sold', 'Sold'), ('cancelled', 'Cancelled'),
        ],
```
1. It's generally a good idea to indent selection values when there is a possibility that the list would be extended in the future.
2. the first part of the tuple is a technical key, it's a bad idea to put spaces inside it them. `offer received -> offer_received`

clbr-odoo commented on estate/security/ir.model.access.csv:
```diff
@@ -0,0 +1,2 @@
+id,name,model_id/id,group_id/id,perm_read,perm_write,perm_create,perm_unlink
+access_estate_user,estate.property,model_estate_property,base.group_user,1,1,1,1
```
1. the second value is the name of the `ir.rule`, not the name of the model. Name it like you want be be specific, name it like the id works.
2. Missing EOL at the end of the file.

clbr-odoo commented on estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,8 @@
+        <menuitem id="estate_model_menu_action" action="estate_model_action"/>
+    </menuitem>
+</menuitem>
+</odoo>
```
EOL.

clbr-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0"?>
+<odoo>
+<record id="estate_model_action" model="ir.actions.act_window">
+    <field name="name">Test properties</field>
```
```suggestion
    <field name="name">Properties</field>
```

clbr-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,8 @@
+    <field name="res_model">estate.property</field>
+    <field name="view_mode">list,form</field>
+</record>
+</odoo>
```
EOL.

clbr-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    'depends': ['base'],
+    'author': "roole",
+    'application': True,
+    'installable': True,
```
You don't really need that key `installable` since `True` is the default value.

clbr-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    'name': "My Awesome module Roole",
+    'version': '1.0',
+    'depends': ['base'],
+    'author': "roole",
```
Fine in this tutorial but know that when developing an Odoo module, we must put "Odoo S.A.".

clbr-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    'installable': True,
+    'category': 'Tutorials',
+    'description': """
+    Description text
```
Take care that triple quote do take spacing into consideration, usually the content must not be indented with the rest.

clbr-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,18 @@
+    'description': """
+    Description text
+    """,
+    'license': 'GPL-3',
```
```suggestion
    'license': 'LGPL-3',
```
We use LGPL-3 as Open-Source/Community license by default. (And `OEEL-1` for Enterprise modules.)

clbr-odoo reviewed (changes requested):
Very good job, just some nitpicks to report 👍

roole-odoo committed:
[IMP] estate: Chapter 6

- Search + filters added
- Enhanced form view
- Added ugly column in the view

roole-odoo committed:
[IMP] estate : Chapter 7

- review remarks applied
- implemented the chapter 7

roole-odoo committed:
[FIX] estate : Fixed Chapter 7

- Added need file to views
- Fixed xmls

roole-odoo committed:
[IMP] estate : Chapter 8

Implemented the chapter 8

roole-odoo committed:
[FIX] estate : check style correction

- correction of the estate_property style PEP8

roole-odoo committed:
[IMP] estate : Chapter 9

- Chapter 9, button implementation

roole-odoo committed:
[IMP] estate : Implementation chapter 10

clbr-odoo commented on estate/models/estate_offer.py:
```diff
@@ -0,0 +1,63 @@
+
+    def action_accept(self):
+        for record in self:
+            for offers in record.property_id.offer_ids:
+                offers.status = "refused"
```
You can actually do this since you're setting all records of the recordset to the same value.
`record.property_id.offer_ids.status = "refused"`

And take care of the plural misused:
```suggestion
            for offer in record.property_id.offer_ids:
                offer.status = "refused"
```

clbr-odoo commented on estate/models/estate_offer.py:
```diff
@@ -0,0 +1,63 @@
+            for offers in record.property_id.offer_ids:
+                offers.status = "refused"
+            record.status = "accepted"
+            record.property_id.state = "offer_accepted"
+            record.property_id.selling_price = record.price
+            record.property_id.buyer_id = record.partner_id
```
```suggestion
            record.property_id.write({
                "state": "offer_accepted",
                "selling_price": record.price,
                "buyer_id": record.partner_id,
            })
```
You can also do this in one call. Tbh the ORM will most likely process all the write in one batch anyway but still, it's a good practice to do this.

clbr-odoo commented on estate/models/estate_offer.py:
```diff
@@ -0,0 +1,63 @@
+
+    @api.depends('validity')
+    def _compute_deadline(self):
+        for record in self:
```
```suggestion
        for estate in self:
```
We more and more try to avoid generic "record" naming in those cases to make it more readable.
This way even when reading a snippet we know what kind of records we're looping on.

clbr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,112 @@
+    @api.depends("offer_ids.price")
+    def _compute_best_price(self):
+        for record in self:
+            if len(record.offer_ids) > 0:
```
```suggestion
            if record.offer_ids:
```
an empty recordset will return a Falsy value when evaluated like this, no need to check the lenght.

clbr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,112 @@
+    def _compute_best_price(self):
+        for record in self:
+            if len(record.offer_ids) > 0:
+                record.best_price = max(record.offer_ids.mapped("price"))
```
You can avoid the if_else and do a one liner like this if you want 😉 
```suggestion
                record.best_price = max(record.offer_ids.mapped("price"), default=0.0)
```

clbr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,112 @@
+                raise UserError("Canceled property can't be sold.")
+            else:
+                record.state = "sold"
+                return True
```
That return should not be in the else clause but at the very end of the method.
But tbh you don't need the else at all.

clbr-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,101 @@
+                    <field name="living_area"/>
+                    <field name="facades"/>
+                    <separator/>
+                    <filter string="Available" name="active" domain="['|', ('state', '=', 'new' ),('state', '=', 'offer received' )]"/>
```
```suggestion
                    <filter string="Available" name="active" domain="[('state', 'in', ('new', 'offer received'))]"/>
```

clbr-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,101 @@
+                    <field name="facades"/>
+                    <separator/>
+                    <filter string="Available" name="active" domain="['|', ('state', '=', 'new' ),('state', '=', 'offer received' )]"/>
+                    <filter string="Postco²de" name="postcode"  context="{'group_by':'postcode'}"/>
```
typo !

clbr-odoo commented on .gitignore:
avoid pushing your personal config 😉

clbr-odoo reviewed (commented):
Very good job, keep up !

roole-odoo committed:
[IMP] estate : Chapter 11

roole-odoo committed:
[FIX] estate : fixed stored attribut name

roole-odoo committed:
[IMP] estate : Chapter 12

roole-odoo committed:
[FIX] estate : fix chapter 12

roole-odoo committed:
[FIX] estate : Chapter 12 fix
